### PR TITLE
[Release] v0.4.1 - Integration / Beta Branch

### PR DESCRIPTION
**Pull request by fbartho**
-- Change PRs are enumerated in the [tracking milestone](https://github.com/taskrabbit/react-native-zendesk-chat/milestone/4)

Changes
- Feature: Ability to register the PushNotification token with Zendesk #77
- Feature: Ability to register the Zendesk AppId (not required for all users) #72
- Android: Fix visitorInfo - email Typo #77 
- Android: Fix visitorInfo + PreChatForm Upstream API Behavior Differences #82 #84
- _Potentially Breaking:_ adds a minimum version of 2.9.0 for ZendeskChatSDK

### CODE REVIEW

**Comment by petchkubb** [comment]
I have tried push notification this version. it's work in android but ios not working. How can I fix it ?

please help.

**Comment by gabrielperales** [comment]
@petchkubb can you try to add this to your `AppDelegate.m`:

```Objective-c
#import <RNCPushNotificationIOS.h>
#import <ChatProvidersSDK/ChatProvidersSDK.h>


- (void)application:(UIApplication *)application didRegisterForRemoteNotificationsWithDeviceToken:(NSData *)deviceToken
{
  [ZDKChat registerPushToken:deviceToken];
  [RNCPushNotificationIOS didRegisterForRemoteNotificationsWithDeviceToken:deviceToken];
}
```

Regards

**Comment by gabrielperales** [comment]
@fbartho I think @petchkubb is right, if I don't add the code I posted before I'm not getting the push notifications. Maybe the token conversion from `NSString` to `NSData` is wrong?

```
RCT_EXPORT_METHOD(registerPushToken:(NSString *)token) {
	dispatch_sync(dispatch_get_main_queue(), ^{
		[ZDKChat registerPushToken:[RCTConvert NSData:token]];
	});
}
```

**Comment by fbartho** [comment]
@gabrielperales / @petchkubb -- Have you added log messages to prove you called the `registerPushToken` API from the JS side of things?

I'm happy to try other ways of converting an NSString to an NSData. `[NSData dataWithString: token]` comes to mind. Have you tested that one?

**Comment by gabrielperales** [comment]
@fbartho No, I haven't. I don't know too much about Objective-C... so when I made it work with the snipped I posted above... I stopped searching for other ways to make it work.

File: android/src/main/java/com/taskrabbit/zendesk/RNZendeskChatModule.java [hunk 1]
```diff
@@ -282,10 +342,38 @@ public void startChat(ReadableMap options) {
 
     @ReactMethod
     public void registerPushToken(String token) {
-        PushNotificationsProvider pushProvider = Chat.INSTANCE.providers().pushNotificationProvider();
+        PushNotificationsProvider pushProvider = Chat.INSTANCE.providers().pushNotificationsProvider();
```
Comment (gabrielperales) [inline android/src/main/java/com/taskrabbit/zendesk/RNZendeskChatModule.java:345]: is this update right? I got the code from the [docs](https://developer.zendesk.com/embeddables/docs/chat-sdk-v-2-for-android/push_notifications) and is says that the method is `pushNotificationProvider()` without the `s`.

```java
PushNotificationsProvider pushProvider = Chat.INSTANCE.providers().pushNotificationProvider();

if (pushProvider != null) {
    pushProvider.registerPushToken("YOUR FIREBASE TOKEN");
}
```

Comment (gabrielperales) [reply android/src/main/java/com/taskrabbit/zendesk/RNZendeskChatModule.java:345]: Ok, I just checked it and it's right.

Comment (petchkubb) [reply android/src/main/java/com/taskrabbit/zendesk/RNZendeskChatModule.java:345]: it's works thank you @gabrielperales

Comment (gabrielperales) [reply android/src/main/java/com/taskrabbit/zendesk/RNZendeskChatModule.java:345]: I'm glad you make it work @petchkubb 👍 . It was working for you in android? I'm not able to receive push notifications from the chat in my android App. Did you configure something I could be missing or something?

**Comment by gabrielperales** [comment]
> @gabrielperales / @petchkubb -- Have you added log messages to prove you called the `registerPushToken` API from the JS side of things?
> 
> I'm happy to try other ways of converting an NSString to an NSData. `[NSData dataWithString: token]` comes to mind. Have you tested that one?

I'm checking that there is another method called [`registerPushTokenString`](https://zendesk.github.io/mobile_sdk_iOS_jazzy_docs/Classes/PushNotificationsProvider.html#/c:@CM@ChatProvidersSDK@objc(cs)ZDKPushNotificationsProvider(im)registerPushTokenString:) which receives the string so we shouldn't have to convert it. I'm going to try with that.

**Comment by gabrielperales** [comment]
@fbartho I changed `registerPushToken` for `registerPushTokenString` and I can confirm that it works :)

**Comment by breadadams** [comment]
@petchkubb did you have to follow any extra steps to get things working on Android?

**Comment by fbartho** [comment]
v0.4.1-beta.7 is ready for anybody watching this PR!

**Pull request merged**